src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,770 @@
+
+            WriteHeader();
+
+            // Buffer the data - don't process it yet
+            _encryptionBuffer?.Write(buffer);
+
+            Debug.WriteLine($"Buffered {buffer.Length} bytes for encryption");
+        }
+
+
+        // Flush all buffered data, encrypt it, and write to base stream
+        private void FlushEncryptionBuffer()
+        {
+            if (_encryptionBuffer != null && _encryptionBuffer.Length > 0)
+            {
+                byte[] data = _encryptionBuffer.ToArray();
+                ProcessBlock(data, 0, data.Length);
+                _baseStream.Write(data);
+                _position += data.Length;
+                _encryptionBuffer.SetLength(0); // Clear the buffer
+            }
+        }
+
+        // Replace the FlushEncryptionBufferAsync method with this version:
+        private async Task FlushEncryptionBufferAsync(CancellationToken cancellationToken)
+        {
+            if (_encryptionBuffer != null && _encryptionBuffer.Length > 0)
+            {
+                byte[] data = _encryptionBuffer.ToArray();
+                ProcessBlock(data, 0, data.Length);
+                await _baseStream.WriteAsync(data, cancellationToken).ConfigureAwait(false);
+                _position += data.Length;
+                _encryptionBuffer.SetLength(0); // Clear the buffer
+            }
+        }
+
+        private int ReadCore(Span<byte> buffer)
+        {
+            ObjectDisposedException.ThrowIf(_disposed, this);
+
+            if (_encrypting)
+                throw new NotSupportedException("Stream is in encryption mode.");
+
+            if (!_headerRead)
+                ReadHeader();
+
+            int bytesToRead = buffer.Length;
+
+            // If we know the total size, ensure we don't read into the HMAC
+            if (_totalStreamSize > 0)
+            {
+                // Calculate the size of the AES overhead
+                int saltSize = _keySizeBits / 16;
+                int headerSize = saltSize + 2; // Salt + Password Verifier
+                const int hmacSize = 10; // 10-byte HMAC
+
+                // The actual encrypted data size is the total minus header and HMAC
+                long encryptedDataSize = _totalStreamSize - headerSize - hmacSize;
+                long remainingData = encryptedDataSize - _bytesReadFromBase;
+
+                if (remainingData <= 0)
+                {
+                    if (!_authCodeValidated)
+                    {
+                        ValidateAuthCode();
+                    }
+                    return 0;
+                }
+
+                bytesToRead = (int)Math.Min(bytesToRead, remainingData);
+            }
+
+            if (bytesToRead == 0)
+            {
+                if (!_authCodeValidated && _totalStreamSize > 0)
+                {
+                    ValidateAuthCode();
+                }
+                return 0;
+            }
+
+            int n = _baseStream.Read(buffer.Slice(0, bytesToRead));
```

Comment:
```suggestion
            int read = _baseStream.Read(buffer.Slice(0, bytesToRead));
```

We generally try to avoid single-letter variable names, with some minor exceptions (like `i` in for loop)